### PR TITLE
enh: series and expr round method

### LINKED ISSUE #302: [Enh]: Support for nw.round()
### We would like to learn about your use case. For example, if this feature is needed  to adopt Narwhals in an open source project, could you please enter the link to it below?

_No response_

### Please describe the purpose of the new feature or describe the problem to solve.

Came up when trying TPC-H query Q11. Issue https://github.com/narwhals-dev/narwhals/issues/282

### Suggest a solution if possible.

_No response_

### If you have tried alternatives, please describe them below.

_No response_

### Would you want to open a pull request?

None

### Additional information that may help us understand your needs.

_No response_

### PR DESCRIPTION
**Pull request by FBruzzesi**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Closes #302 

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added 
- [x] Documented the changes

## If you have comments or can explain your changes, please do so below.

I just realized polars does not have a `DataFrame.round`, I guess `df.select(pl.all().round())` would do the trick :)

### CODE REVIEW

File: narwhals/_pandas_like/expr.py [hunk 1]
```diff
@@ -294,6 +294,9 @@ def head(self, n: int) -> Self:
     def tail(self, n: int) -> Self:
         return reuse_series_implementation(self, "tail", n)
 
+    def round(self: Self, decimals: int = 0) -> Self:
```
Comment (MarcoGorelli) [inline narwhals/_pandas_like/expr.py:297]: i should add a pre-commit rule here, but i think we should avoid default args in _pandas_like
```suggestion
    def round(self: Self, decimals: int) -> Self:
```

File: narwhals/_pandas_like/series.py [hunk 2]
```diff
@@ -500,6 +500,9 @@ def head(self: Self, n: int) -> Self:
     def tail(self: Self, n: int) -> Self:
         return self._from_series(self._series.tail(n))
 
+    def round(self: Self, decimals: int = 0) -> Self:
```
Comment (MarcoGorelli) [inline narwhals/_pandas_like/series.py:503]:
```suggestion
    def round(self: Self, decimals: int) -> Self:
```

File: narwhals/expression.py [hunk 3]
```diff
@@ -1519,6 +1519,49 @@ def tail(self, n: int = 10) -> Expr:
 
         return self.__class__(lambda plx: self._call(plx).tail(n))
 
+    def round(self, decimals: int = 0) -> Expr:
+        r"""
+        Round underlying floating point data by `decimals` digits.
```
Comment (MarcoGorelli) [inline narwhals/expression.py:1524]: i think pandas and polars may differ in how they round from midway (`.5`, `1.5`), shall we add a note about it?

File: tests/test_common.py [hunk 4]
```diff
@@ -765,3 +765,19 @@ def test_item_value_error(
 ) -> None:
     with pytest.raises(ValueError, match=err_msg):
         nw.from_native(df_raw, eager_only=True).item(row, column)
+
+
+@pytest.mark.parametrize("constructor", [pd.DataFrame, pl.DataFrame])
+@pytest.mark.parametrize("decimals", [0, 1, 2])
+def test_round(constructor: Any, decimals: int) -> None:
```
Comment (MarcoGorelli) [inline tests/test_common.py:772]: shall we move this to `tests/expr/round_test`? i think going forwards it's good to aim for one test per method

**Review by MarcoGorelli — COMMENTED** [review]
awesome! just some very minor comments

File: narwhals/expression.py (hunk 3 above)
Comment (FBruzzesi) [reply narwhals/expression.py:1524]: Do you know if the difference only applies in the range `(-.5, .5)`? 

> shall we add a note about it?
If we know how to correct it manually, I think we should. Steering pandas behavior to polars' one

Comment (MarcoGorelli) [reply narwhals/expression.py:1524]: i think pandas rounds to the nearest even, polars towards zero? something like that

not sure about how to correct it reliably - maybe we can just document it for now?

Comment (FBruzzesi) [reply narwhals/expression.py:1524]: I added a note, but I would rather have a strategy to address cases like this one in the future 🤔

File: narwhals/expression.py [hunk 5]
```diff
@@ -1519,6 +1519,58 @@ def tail(self, n: int = 10) -> Expr:
...
+        Arguments:
+            decimals: Number of decimals to round by.
+
+        Notes:
+            For values exactly halfway between rounded decimal values pandas and Polars behave differently.
+
+            pandas rounds to the nearest even value (e.g. -0.5 and 0.5 round to 0.0, 1.5 and 2.5 round to 2.0, 3.5 and
+            4.5 to 4.0, etc..).
+
+            Polars rounds towards the larger integer (e.g. -0.5 to -1.0, 0.5 to 1.0, 1.5 to 2.0, 2.5 to 3.0, etc..).
```
Comment (MarcoGorelli) [inline narwhals/expression.py:1535]:
```suggestion
            Polars rounds away from 0 (e.g. -0.5 to -1.0, 0.5 to 1.0, 1.5 to 2.0, 2.5 to 3.0, etc..).
```

File: narwhals/series.py [hunk 6]
```diff
@@ -1707,6 +1707,54 @@ def tail(self: Self, n: int = 10) -> Self:
...
+        Arguments
+            decimals: Number of decimals to round by.
+
+        Notes:
+            For values exactly halfway between rounded decimal values pandas and Polars behave differently.
+
+            pandas rounds to the nearest even value (e.g. -0.5 and 0.5 round to 0.0, 1.5 and 2.5 round to 2.0, 3.5 and
+            4.5 to 4.0, etc..).
+
+            Polars rounds towards the larger integer (e.g. -0.5 to -1.0, 0.5 to 1.0, 1.5 to 2.0, 2.5 to 3.0, etc..).
```
Comment (MarcoGorelli) [inline narwhals/series.py:1723]:
```suggestion
            Polars rounds away from 0 (e.g. -0.5 to -1.0, 0.5 to 1.0, 1.5 to 2.0, 2.5 to 3.0, etc..).
```

**Review by MarcoGorelli — APPROVED** [review]
awesome, thanks @FBruzzesi ! feel free to merge

I just made a tiny edit, as I think "away from 0" is more accurate than "towards the larger integer", as 0 is larger than -1 even though its magnitude is smaller. nitpicking here, sorry 😛 I think the examples you added explain it perfectly - the Narwhals docs may even end up being a useful reference of pandas vs Polars differences 😄

**Pull request merged**